torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -231,23 +233,82 @@ class ValidateAllocation : private kir::IrVisitor {
   std::vector<std::vector<const Allocate*>> live_allocations_;
 };
 
+class MemoizeAllocations : kir::IrVisitor {
+ public:
+  static void memoizeAllocations(Kernel* kernel, ExprBuilderVisitor& visitor) {
+    MemoizeAllocations(visitor, kernel, kernel->summary());
+  }
+
+ private:
+  MemoizeAllocations(
+      ExprBuilderVisitor& builder_visitor,
+      const Kernel* kernel,
+      KernelSummary& summary)
+      : summary_(&summary) {
+    TORCH_CHECK(summary_->memoized_dag == nullptr);
+    builder_visitor_ =
+        &builder_visitor; //.reset(new ExprBuilderVisitor(builder_, *kernel));
+
+    for (const auto& ir_node : kernel->irNodes())
+      ir_node->accept(this);
+  }
+
+  void memoizeGlobalAllocationShape(const kir::Allocate* allocate) {
+    std::vector<glfdc::ExprEvaluator> shapes;
+
+    for (auto shape_expr : allocate->shape()) {
+      auto expr = builder_visitor_->build(shape_expr);
+      TORCH_CHECK(expr.has_value());
```

Comment:
Is it guaranteed to has a value? If not, please add an error message to the assertion.

Reply:
Return type of build method doesn't need to be optional. Moreover it can return expression in memoized form to avoid additional steps.

Reply:
This shouldn't be concern in current version as build return result is not any longer c10::optional<Expr>.